metadata-ingestion/tests/unit/sql_parsing/test_sql_aggregator.py

```diff
@@ -1025,3 +1026,411 @@ def test_sql_aggreator_close_cleans_tmp(tmp_path):
         assert len(os.listdir(tmp_path)) > 0
         aggregator.close()
         assert len(os.listdir(tmp_path)) == 0
+
+
+@freeze_time(FROZEN_TIME)
+def test_diamond_problem(pytestconfig: pytest.Config, tmp_path: pathlib.Path) -> None:
+    aggregator = SqlParsingAggregator(
+        platform="snowflake",
+        generate_lineage=True,
+        generate_usage_statistics=False,
+        generate_operations=False,
+        is_temp_table=lambda x: x.lower()
+        in [
+            "diamond_problem.dummy_test.diamond_problem.t1",
+            "diamond_problem.dummy_test.diamond_problem.t2",
+            "diamond_problem.dummy_test.diamond_problem.t3",
+            "diamond_problem.dummy_test.diamond_problem.t4",
+        ],
+    )
+
+    aggregator._schema_resolver.add_raw_schema_info(
+        DatasetUrn(
+            "snowflake", "diamond_problem.dummy_test.diamond_problem.diamond_source1"
+        ).urn(),
+        {"col_a": "int", "col_b": "int", "col_c": "int"},
+    )
+
+    aggregator._schema_resolver.add_raw_schema_info(
+        DatasetUrn(
+            "snowflake",
+            "diamond_problem.dummy_test.diamond_problem.diamond_destination",
+        ).urn(),
+        {"col_a": "int", "col_b": "int", "col_c": "int"},
+    )
+
+    aggregator.add_preparsed_query(
+        PreparsedQuery(
+            query_id="62375ec02d44aa1fbaa8213ddd8d9e33d5ed80d93bd0e51942aa18f5643979d2",
+            query_text="CREATE TEMPORARY TABLE t1 as select * from diamond_source1;",
+            upstreams=[
+                DatasetUrn(
+                    "snowflake",
+                    "diamond_problem.dummy_test.diamond_problem.diamond_source1",
+                ).urn()
+            ],
+            downstream=DatasetUrn(
+                "snowflake", "diamond_problem.dummy_test.diamond_problem.t1"
+            ).urn(),
```

Comment:
is there a reason we're not just using add_observed_query here? that'd make the test less verbose

Reply:
Snowflake source will generate `PreparsedQuery` for the queries triggering that diamond case - so we are getting upstreams/downstreams from the audit log instead of parsing those queries. Do you still want me to change it to `ObservedQuery`? This will be less redundant in the code of course, will not exactly test our original case (Snowflake ingestion), on the other hand it should be effectively the same.